common/lsp/lsp-file-utils_test.cc

```diff
@@ -0,0 +1,83 @@
+// Copyright 2023 The Verible Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#include "common/lsp/lsp-file-utils.h"
+
+#include "gtest/gtest.h"
+
+TEST(URIDecodingTest, SimplePathDecoding) {
+  std::string exemplar = "/home/user/test-project/add.sv";
+  std::string result =
+      verible::lsp::LSPUriToPath("file:///home/user/test-project/add.sv");
+
+  ASSERT_EQ(exemplar, result);
+}
+
+TEST(URIDecodingTest, WindowsPathDecoding) {
+  std::string exemplar = "c:/Users/user/test-project/add.sv";
+  std::string result =
+      verible::lsp::LSPUriToPath("file:///c%3A/Users/user/test-project/add.sv");
+
+  ASSERT_EQ(exemplar, result);
+}
+
+TEST(URIDecodingtest, PathWithSpacesDecoding) {
+  std::string exemplar = "/home/user/test project/my module.sv";
+  std::string result = verible::lsp::LSPUriToPath(
+      "file:///home/user/test%20project/my%20module.sv");
+
+  ASSERT_EQ(exemplar, result);
+}
+
+TEST(URIDecodingtest, PathWithConsecutiveEscapes) {
+  std::string exemplar = "/home/user/test-project/my    module.sv";
+  std::string result = verible::lsp::LSPUriToPath(
+      "file:///home/user/test-project/my%20%20%20%20module.sv");
+
+  ASSERT_EQ(exemplar, result);
+}
+
+TEST(URIEncodingTest, SimplePathEncoding) {
+  std::string exemplar = "file:///home/user/test-project/add.sv";
+  std::string result =
+      verible::lsp::PathToLSPUri("/home/user/test-project/add.sv");
+
+  ASSERT_EQ(exemplar, result);
+}
+
+TEST(URIEncodingTest, WindowsPathEncoding) {
+  std::string exemplar = "file:///c%3a/Users/user/test-project/add.sv";
+  std::string result =
+      verible::lsp::PathToLSPUri("c:/Users/user/test-project/add.sv");
+
+  ASSERT_EQ(exemplar, result);
+}
+
+TEST(URIEncodingTest, PathWithSpacesEncoding) {
+  std::string exemplar = "file:///home/user/test%20project/my%20module.sv";
+  std::string result =
+      verible::lsp::PathToLSPUri("/home/user/test project/my module.sv");
+
+  ASSERT_EQ(exemplar, result);
+}
+
+TEST(URIEncodingTest, PathWithConsecutiveEscapes) {
+  std::string exemplar =
+      "file:///home/user/test-project/my%20%20%20%20module.sv";
+  std::string result =
+      verible::lsp::PathToLSPUri("/home/user/test-project/my    module.sv");
+
+  ASSERT_EQ(exemplar, result);
+}
```

Comment:
Also add tests with broken escape sequences `/foo/bar/%xyz` ... they should come out as-is.